tests/cpp_unit_tests/test_tap_position_optimizer.cpp

```diff
@@ -917,122 +1041,133 @@ TEST_CASE("Test Tap position optimizer") {
-                    // For a tap_size of 0.1 and tap_nom of 0, tap_pos_relative_a = 0.1 * (tap_pos_a - 0)
-                    auto const relative_tap_a = static_cast<double>(state_a.tap_pos) * 0.1;
-                    return static_cast<DoubleComplex>(1.0 / (1.0 + relative_tap_a));
-                };
-
-                state_b.u_pu = [&state_a, &regulator_a, &state_b, &regulator_b](ControlSide side) {
-                    CHECK(side == regulator_b.control_side());
+                SUBCASE("multiple transformers with control function based on ranking") {
+                    state_a.rank = 0;
+                    state_b.rank = 1;
+                    state_a.tap_min = -5;
+                    state_a.tap_max = 5;
+                    state_b.tap_min = -5;
+                    state_b.tap_max = 5;
+
+                    state_a.u_pu = [&state_a, &regulator_a](ControlSide side) {
+                        CHECK(side == regulator_a.control_side());
+
+                        // u_2a = f(tap_pos_a) when rank is 0
+                        // u_2a = (u_1a * n_1) / (1.0 + relative_tap_pos_a)
+                        // consider u_1a = n_1 = 1.0
+                        // For a tap_size of 0.1 and tap_nom of 0, tap_pos_relative_a = 0.1 * (tap_pos_a - 0)
+                        auto const relative_tap_a = static_cast<double>(state_a.tap_pos) * 0.1;
+                        if (state_a.tap_side == regulator_a.control_side()) {
+                            return static_cast<DoubleComplex>(1.0 + relative_tap_a);
+                        }
+                        return static_cast<DoubleComplex>(1.0 / (1.0 + relative_tap_a));
+                    };
 
-                    // u_2b = f(tap_pos_a, tap_pos_b) when rank is 1
-                    // u_2b = (u_1b * n_2) / (1.0 + relative_tap_pos_b)
-                    // consider n_2 = 1. Also u_1a = u_2b
-                    // For a tap_size of 0.1 and tap_nom of 0, tap_pos_relative_b = 0.1 * (tap_pos_b - 0)
-                    auto const relative_tap_b = static_cast<double>(state_b.tap_pos) * 0.1;
-                    return state_a.u_pu(regulator_a.control_side()) / (1.0 + relative_tap_b);
-                };
+                    state_b.u_pu = [&state_a, &regulator_a, &state_b, &regulator_b](ControlSide side) {
+                        CHECK(side == regulator_b.control_side());
+
+                        // u_2b = f(tap_pos_a, tap_pos_b) when rank is 1
+                        // u_2b = (u_1b * n_2) / (1.0 + relative_tap_pos_b)
+                        // consider n_2 = 1. Also u_1a = u_2b
+                        // For a tap_size of 0.1 and tap_nom of 0, tap_pos_relative_b = 0.1 * (tap_pos_b - 0)
+                        auto const relative_tap_b = static_cast<double>(state_b.tap_pos) * 0.1;
+                        if (state_b.tap_side == regulator_b.control_side()) {
+                            return (1.0 + relative_tap_b) / state_a.u_pu(regulator_a.control_side());
```

Comment:
i don't think this is correct unit-wise. I think it should just be
```suggestion
                            return (1.0 + relative_tap_b) * state_a.u_pu(regulator_a.control_side());
```

that may also make the rest of the logic simpler